packages/liveblocks-core/src/lib/objectToQuery.ts

```diff
@@ -0,0 +1,222 @@
+/**
+ * Converts an object to a query string
+ * Example:
+ * ```ts
+ * const query = objectToQuery({
+      resolved: true,
+      metadata: {
+        status: "open",
+        priority: 3,
+        org: {
+          startsWith: "liveblocks:",
+        },
+      },
+});
+
+console.log(query);
+// resolved:true AND metadata["status"]:open AND metadata["priority"]:3 AND metadata["org"]^"liveblocks:"
+
+ * ```
+ *
+ *
+ */
+
+type FilterValue = string | number | boolean | { startsWith: string };
+
+type Filter = NumberFilter | StringFilter | BooleanFilter;
+
+type NumberFilter = {
+  key: string;
+  operator: ":";
+  value: number;
+};
+
+type StringFilter = {
+  key: string;
+  operator: ":" | "^";
+  value: string;
+};
+
+type BooleanFilter = {
+  key: string;
+  operator: ":";
+  value: boolean;
+};
+
+/**
+ * Converts an object to a query string
+ * @example
+ * ```ts
+ * const query = objectToQuery({
+ *  metadata: {
+ *    status: "open",
+ *    priority: 3,
+ *    org: {
+ *      startsWith: "liveblocks:",
+ *    },
+ *  },
+ * });
+ * console.log(query);
+ * // metadata["status"]:"open" AND metadata["priority"]:3 AND metadata["org"]^"liveblocks:"
+ * ```
+ * @example
+ * ```ts
+ * const query = objectToQuery({
+ *  resolved: true,
+ *  roomId: {
+ *   startsWith: "engineering:",
+ *  },
+ * });
+ * console.log(query);
+ * // resolved:true AND roomId^"engineering:"
+ * ```
+ *
+ */
+export function objectToQuery(obj: {
+  [key: string]: FilterValue | { [key: string]: FilterValue };
```

Comment:
You'll need to account for explicit `undefined`s here too at the TypeScript level. We typically would consider these values to be equal:

```ts
{
  foo: "bar"
}
```

```ts
{
  foo: "bar",
  qux: undefined,
}
```

…but TypeScript knows they're not equal. I think in a function like this, it's fine to accept values containing explicit-`undefined`s and simply ignoring/skip over those keys. Accepting those values makes it much more pragmatic to use this API, as you can use it to conditionally include fields or not, for example:

```ts
{
  foo: myCondition ? "bar" : undefined
}
```